packages/docusaurus-theme-common/src/contexts/colorMode.tsx

```diff
@@ -65,10 +65,34 @@ function useContextValue(): ContextValue {
     getInitialColorMode(defaultMode),
   );
 
-  const setColorMode = useCallback((newColorMode: ColorMode) => {
-    setColorModeState(newColorMode);
-    storeColorMode(newColorMode);
-  }, []);
+  useEffect(() => {
+    // A site is deployed without disableSwitch
+    // => User visits the site and has a persisted value
+    // => Site later enabled disableSwitch
+    // => Clear the previously stored value to apply the site's setting
+    if (disableSwitch) {
+      ColorModeStorage.del();
+    }
+  }, [disableSwitch]);
+
+  const setColorMode = useCallback(
+    (newColorMode: ColorMode | null, persist = true) => {
+      if (newColorMode) {
+        setColorModeState(newColorMode);
+        if (persist) {
+          storeColorMode(newColorMode);
+        }
+      } else if (respectPrefersColorScheme) {
+        setColorMode(
+          window.matchMedia('(prefers-color-scheme: dark)').matches
+            ? ColorModes.dark
+            : ColorModes.light,
+        );
+        ColorModeStorage.del();
+      }
```

Comment:
New API: `setColorMode(null)`, when `respectPrefersColorScheme`, will reset the theme to the OS default and delete the storage. This should enable users to do meaningful things.

Reply:
Should it fall back to `defaultMode` then? I'm fine with that, didn't think about it